Call ClearScriptBlockCache to prevent OutOfMemoryExceptions ClearScriptBlockCache

mwrock opened the pull request:
Powershell caches ScritBlocks in a dictionary keyed on the script block text. Thats just great unless the script is super large and called a gillion times like we might do. In such a case it will saturate the Large Object Heap and lead to Out Of Memory exceptions for large files or folders. So we call the internal method
ClearScriptBlockCache to clear it.

mwrock commented:
See https://github.com/PowerShell/PowerShell/blob/0cd1a0ff48c5ebe4723b454c7c3fd59e4b2c87de/src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs#L526

mwrock commented:
Watching the `private bytes` count of the remote powershell process, the count seemed to grow ~3k on each upload chunk and would climb steadily until it hit the `MaxMemoryPerShellMb` threshold (default 1024 on 2012R2).

I initially assumed this was somewhere in the `filestream.Write` but nothing I tried really moved the needle. Adding a `GC.Collect()` helped but not significantly. Out of curiosity I completely remover the file write and observed the memory rising just by setting the bytes variable.

Attaching a memory profiler I could see the large `$byte` strings hanging out in the Large Object Heap in a `Dictionary<Tuple<String, String>, ScriptBlock>` and rooted to a `Object[]` indicating some kind of static reference. I eventually stumbled on the internal `_ScriptCache` noted in the comment above.

With this change, the process remains at about 35MB throughout the entire upload and there is no need to explicitly invoke the garbage collector.

sneal commented on lib/winrm-fs/core/file_transporter.rb:
```diff
@@ -443,6 +443,16 @@ def stream_upload(input_io, dest)
 
         def stream_command(encoded_bytes)
           <<-EOS
+            # Powershell caches ScritBlocks in a dictionary
```
typo in 'ScritBlocks'

sneal replied:
Probably better to move the comment outside the powershell script to avoid uploading it?

sneal commented on lib/winrm-fs/core/file_transporter.rb:
```diff
@@ -443,6 +443,16 @@ def stream_upload(input_io, dest)
+            # Large Object Heap and lead to Out Of Memory exceptions
+            # for large files or folders. So we call the internal method
+            # ClearScriptBlockCache to clear it.
+            $BindingFlags= [Reflection.BindingFlags] "NonPublic,Static"
```
For consistency shouldn't `$BindingFlags` be camel cased?

mwrock commented:
Good catches. Fixed camelcasing and moved method lookup out of the streaming block and into the same block where we initialize the FileStream so we don't make a relatively expensive reflection lookup on every chunk when they all yield the same data.

I kept the comment inline and think its fine to upload just as long as we dont do it every time eating into our buffer.
